lib/ansible/plugins/connection/vmware_tools.py

```diff
@@ -0,0 +1,297 @@
+        """Reset the connection."""
+        super(Connection, self).reset()
+
+        self.close()
+        self._connect()
+
+    def create_temporary_file_in_guest(self, prefix="", suffix=""):
+        """Create a temporary file in the VM."""
+        return self.fileManager.CreateTemporaryFileInGuest(vm=self.vm, auth=self.vm_auth, prefix=prefix, suffix=suffix)
+
+    def _get_program_spec_program_path_and_arguments(self, cmd):
+        program_path = self._play_context.executable
+        arguments = re.sub(r"^{}\s*".format(program_path), "", cmd)
+
+        return program_path, arguments
+
+    def _get_guest_program_spec(self, cmd, stdout, stderr):
+        guest_program_spec = vim.GuestProgramSpec()
+
+        program_path, arguments = self._get_program_spec_program_path_and_arguments(cmd)
+
+        arguments += " 1> {} 2> {}".format(stdout, stderr)
+
+        guest_program_spec.programPath = program_path
+        guest_program_spec.arguments = arguments
+
+        return guest_program_spec
+
+    def get_pid_info(self, pid):
+        """Return pid status."""
+        processes = self.processManager.ListProcessesInGuest(vm=self.vm, auth=self.vm_auth, pids=[pid])
+        return processes[0]
+
+    def _fetch_file_from_vm(self, guestFilePath):
+        fileTransferInformation = self.fileManager.InitiateFileTransferFromGuest(vm=self.vm, auth=self.vm_auth, guestFilePath=guestFilePath)
+        response = requests.get(fileTransferInformation.url, verify=self.connection_verify_ssl, stream=True)
+
+        if response.status_code != 200:
+            raise AnsibleError("Failed to fetch file")
+
+        return response
+
+    def delete_file_in_guest(self, filePath):
+        """Delete file from VM."""
+        self.fileManager.DeleteFileInGuest(vm=self.vm, auth=self.vm_auth, filePath=filePath)
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        """Execute command."""
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        stdout = self.create_temporary_file_in_guest(suffix=".stdout")
+        stderr = self.create_temporary_file_in_guest(suffix=".stderr")
+
+        guest_program_spec = self._get_guest_program_spec(cmd, stdout, stderr)
+
+        try:
+            pid = self.processManager.StartProgramInGuest(vm=self.vm, auth=self.vm_auth, spec=guest_program_spec)
+        except vim.fault.FileNotFound as e:
+            raise AnsibleError("StartProgramInGuest Error: {}".format(e.msg))
+
+        pid_info = self.get_pid_info(pid)
+
+        while pid_info.endTime is None:
+            sleep(self.get_option("exec_command_sleep_interval"))
+            pid_info = self.get_pid_info(pid)
+
+        stdout_response = self._fetch_file_from_vm(stdout)
+        self.delete_file_in_guest(stdout)
+
+        stderr_response = self._fetch_file_from_vm(stderr)
+        self.delete_file_in_guest(stderr)
+
+        return pid_info.exitCode, stdout_response.text, stderr_response.text
```

Comment:
Does anyone know if `stdout_response.text` & `stderr_response.text` needs to be wrapped in `to_text`? I was just looking at [string-encoding](https://docs.ansible.com/ansible/devel/dev_guide/developing_plugins.html#string-encoding) in the dev guide, but I don't see a lot of other connection plugins doing that.

Reply:
I think it is not required since it is not being used in code and just returning to user via playbook register.